File: go/05-slices-maps/main_test.go
package main

import "testing"

func TestAddAndGetTask(t *testing.T) {
	r := NewRobot("R2D2")
	r.AddTask("scan", "Scan surroundings")

	desc, ok := r.GetTask("scan")
	if !ok {
		t.Fatal("expected task 'scan' to exist")
	}
	if desc != "Scan surroundings" {
		t.Errorf("got %q, want %q", desc, "Scan surroundings")
	}
}

func TestGetTaskMissing(t *testing.T) {
	r := NewRobot("R2D2")
	_, ok := r.GetTask("missing")
	if ok {
		t.Error("expected missing task to return false")
	}
}

func TestRecordAndAverageBattery(t *testing.T) {
	r := NewRobot("R2D2")
	r.Battery = 90
	r.RecordBattery()
	r.Battery = 80
	r.RecordBattery()
	r.Battery = 70
	r.RecordBattery()

	got := r.AverageBattery()
	if got != 80.0 {
		t.Errorf("AverageBattery() = %f, want 80.0", got)
	}
}

func TestAverageBatteryEmpty(t *testing.T) {
	r := NewRobot("R2D2")
	if got := r.AverageBattery(); got != 0 {
		t.Errorf("AverageBattery() with no history = %f, want 0", got)
	}
}

File: rust/05-vecs-hashmaps/tests/main_test.rs
use lesson_05_vecs_hashmaps::Robot;

#[test]
fn test_add_and_get_task() {
    let mut r = Robot::new("R2D2");
    r.add_task("scan", "Scan surroundings");
    assert_eq!(r.get_task("scan"), Some(&"Scan surroundings".to_string()));
}

#[test]
fn test_get_task_missing() {
    let r = Robot::new("R2D2");
    assert_eq!(r.get_task("missing"), None);
}

#[test]
fn test_record_and_average_battery() {
    let mut r = Robot::new("R2D2");
    r.battery = 90; r.record_battery();
    r.battery = 80; r.record_battery();
    r.battery = 70; r.record_battery();
    assert_eq!(r.average_battery(), 80.0);
}

#[test]
fn test_average_battery_empty() {
    let r = Robot::new("R2D2");
    assert_eq!(r.average_battery(), 0.0);
}
